hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/AbstractContainerReportHandler.java

```diff
@@ -103,14 +108,44 @@ private void updateContainerStats(final ContainerID containerId,
         containerInfo.updateSequenceId(
             replicaProto.getBlockCommitSequenceId());
       }
+      List<ContainerReplica> otherReplicas =
+          getOtherReplicas(containerId, datanodeDetails);
```

Comment:
I think we can skip filtering the replicas for "others" (and omit the code for that), since the resulting count/bytes values are the min/max of all replicas, including this one, either way.  If we loop over all replicas, the iteration for the current replica will be a "no-op", not change the result.

Reply:
I think we still do need to filter out the current replica. The reason is that at the point the stats are updated, the newly reported replica has not yet updated the previous values stored for it in SCM memory. This is due to the order of operations here, in AbstractContainerReportHandler:

```
  protected void processContainerReplica(final DatanodeDetails datanodeDetails,
                               final ContainerReplicaProto replicaProto)
      throws IOException {
    final ContainerID containerId = ContainerID
        .valueof(replicaProto.getContainerID());

    if (logger.isDebugEnabled()) {
      logger.debug("Processing replica of container {} from datanode {}",
          containerId, datanodeDetails);
    }
    // Synchronized block should be replaced by container lock,
    // once we have introduced lock inside ContainerInfo.
    synchronized (containerManager.getContainer(containerId)) {
      updateContainerStats(datanodeDetails, containerId, replicaProto);
      updateContainerState(datanodeDetails, containerId, replicaProto);
      updateContainerReplica(datanodeDetails, containerId, replicaProto);
    }
  }
```

The in-memory values for the reported replicas are not updated until the updateContainerReplica(...) call, but we updateContainerStats(...) before that.

For example, lets say we have an CLOSED container, so its the max of the values we use.

Existing replicas have values 4, 4, 5, so the value used is 5.

The new report changes the 5 to 4, so now we would have:

4, 4, 5 (stale value, will be updated soon), 4(new value of stale value).

If we don't filter out the stale replica, then the value will be wrong and left as 5 until the next container report is processed, when it would change to 4.

Does that make sense?

Reply:
Good point, I missed that.  Thanks for the explanation.